File: 3250-find-the-count-of-monotonic-pairs-i/3250-find-the-count-of-monotonic-pairs-i.cpp
class Solution {
public:
    int mod=1e9+7;
    int countOfPairs(vector<int>& nums) {
        int n=nums.size();
        vector<vector<int>> cache(n+1,vector<int>(51,-1)); 
        int ans=solve(nums,0,0,cache);
        return cache[0][0];
    }

    int solve(vector<int>& nums,int index,int arr1B,vector<vector<int>> & cache){
        if(index>=nums.size()){
            return 1;
        }

        int arr2B=50;
        if(index>0){
            arr2B=nums[index-1]-arr1B;
        }

        if(cache[index][arr1B]!=-1){
            return cache[index][arr1B];
        }

        int ans=0;
        for(int newArr1B=arr1B;newArr1B<=nums[index];newArr1B++){
            int newArr2B=nums[index]-newArr1B;
            if(newArr2B<=arr2B){
                ans+=solve(nums,index+1,newArr1B,cache)%mod;
                ans=ans%mod;
            }
        }
        return cache[index][arr1B]=ans;
    }
};


/*


2   3   2

0-2 

2-0 


arr1 is increasing 


arr2 is decreasing

arr1[i]+arr2[i]==nums[i]


*/

File: 3250-find-the-count-of-monotonic-pairs-i/3250-find-the-count-of-monotonic-pairs-i.java
class Solution {
    HashMap<String,Long> memo = new HashMap<>();
    public long helper(int i,int nums[],int s,int e){
         if(i==nums.length){
            return 1;
        }
        String key = s+"-"+e+"-"+i;
        if(memo.containsKey(key)){
            return memo.get(key);
        }
        long ans = 0;
        int j= s;
        int k = Math.min(e,nums[i]);
        while(j<=nums[i] && k>=0){
            int sum = j+k;
            if(sum==nums[i]){
                ans = (ans + helper(i+1,nums,j,k))%1000000007;
                j++;
                k--;
            } else if(sum<nums[i]){
                j++;
            }else {
                k--;
            }
        }
        memo.put(key,ans);
        return ans;
    }
    public int countOfPairs(int[] nums) {
        long ans = helper(0,nums,0,50);
        return (int)ans;
    }
}